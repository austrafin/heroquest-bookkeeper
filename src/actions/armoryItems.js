export const SET = "armoryItems/SET";
export const UPDATE = "armoryItems/UPDATE";

export function setArmoryItem(value, key, id) {
  return {
    type: SET,
    key: key,
    id: id,
    value: value,
  };
}

export function updateArmoryItem(id) {
  return {
    type: UPDATE,
    id: id,
  };
}
