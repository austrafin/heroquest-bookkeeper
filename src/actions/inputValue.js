export const SET = "inputValue/SET";

export function setInputValue(value, status) {
  return {
    type: SET,
    status: status,
    value: value
  };
}
