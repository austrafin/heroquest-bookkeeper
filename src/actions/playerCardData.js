export const INCREMENT = "statusPoints/INCREMENT";
export const DECREMENT = "statusPoints/DECREMENT";
export const INITIALISE = "statusPoints/INITIALISE";

export function initialise(data) {
  return {
    type: INITIALISE,
    data: data,
  };
}

export function increment(value, label, cardId) {
  return {
    type: INCREMENT,
    incrementValue: value,
    label: label,
    cardId: cardId,
  };
}

export function decrement(value, label, cardId) {
  return {
    type: DECREMENT,
    incrementValue: value,
    label: label,
    cardId: cardId,
  };
}
