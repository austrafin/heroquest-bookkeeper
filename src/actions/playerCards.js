export const ADD = "playerCards/ADD";
export const ADD_AFTER = "playerCards/ADD_AFTER";
export const LOAD = "playerCards/LOAD";
export const CARDS_LOADED = "playerCards/CARDS_LOADED";

export function addPlayerCard(values) {
  return {
    type: ADD,
    values: values,
  };
}

export function loadPlayerCards() {
  return {
    type: LOAD,
  };
}
