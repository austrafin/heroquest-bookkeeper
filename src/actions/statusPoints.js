export const INCREMENT = "statusPoints/INCREMENT";
export const DECREMENT = "statusPoints/DECREMENT";
export const INITIALISE = "statusPoints/INITIALISE";

export function initialise(data) {
  return {
    type: INITIALISE,
    data: data
  };
}

export function increment(value, label) {
  return {
    type: INCREMENT,
    incrementValue: value,
    label: label
  };
}

export function decrement(value, label) {
  return {
    type: DECREMENT,
    incrementValue: value,
    label: label
  };
}
