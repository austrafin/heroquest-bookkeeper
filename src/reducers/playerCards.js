import {
  ADD,
  ADD_AFTER,
  CARDS_LOADED,
  SET_VALUE,
  INCREMENT,
  DECREMENT,
  INITIALISE,
  SET_SELECTED_IMAGE,
  UPDATE_BASE_VALUES,
} from "../actions/playerCards";

function difference(state, action, decrement) {
  let incrementValue = action.incrementValue;

  if (decrement) {
    incrementValue *= -1;
  }

  if (action.cardId in state.cardData) {
    return {
      ...state,
      cardData: {
        ...state.cardData,
        [action.cardId]: {
          ...state.cardData[action.cardId],
          [action.label]:
            state.cardData[action.cardId][action.label] + incrementValue,
        },
      },
    };
  }
  return {
    ...state,
    cardData: {
      ...state.cardData,
      [action.cardId]: {
        ...state.cardData[action.cardId],
        [action.label]: 0 + incrementValue,
      },
    },
  };
}

export default (state = {}, action) => {
  switch (action.type) {
    case SET_VALUE:
      if (action.cardId in state.cardData) {
        return {
          ...state,
          cardData: {
            ...state.cardData,
            [action.cardId]: {
              ...state.cardData[action.cardId],
              [action.label]: action.value,
            },
          },
        };
      }
      return state;
    case INCREMENT:
      return difference(state, action, false);
    case DECREMENT:
      return difference(state, action, true);
    case INITIALISE:
      return {
        ...state,
        cardData: {
          ...action.data,
        },
      };
    case ADD:
      return { ...state, newCardUploading: true };
    case ADD_AFTER:
      return { ...state, newCardUploading: false };
    case CARDS_LOADED:
      return { ...state, cardsLoaded: action.value };
    case SET_SELECTED_IMAGE:
      return {
        ...state,
        cardData: {
          ...state.cardData,
          [action.cardId]: {
            ...state.cardData[action.cardId],
            selectedImageFile: action.selectedFile,
          },
        },
      };
    case UPDATE_BASE_VALUES:
      console.log(action);
      return {
        ...state,
        cardData: {
          ...state.cardData,
          [action.cardId]: {
            ...state.cardData[action.cardId],
            baseBodyPoints: action.values.baseBodyPoints,
            baseMindPoints: action.values.baseMindPoints,
            baseMeleePoints: action.values.baseMeleePoints,
            baseRangedPoints: action.values.baseRangedPoints,
            baseDiagonalPoints: action.values.baseDiagonalPoints,
            baseDefencePoints: action.values.baseDefencePoints,
            baseMovementPoints: action.values.baseMovementPoints,
          },
        },
      };
    default:
      return state;
  }
};
