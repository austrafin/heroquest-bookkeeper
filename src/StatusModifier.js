import React from "react";
import CardNumberInput from "./CardNumberInput";
import { Button, ButtonGroup, Grid } from "@material-ui/core";
import { useDispatch, useSelector } from "react-redux";
import { increment, decrement } from "./actions/playerCards";

export default (props) => {
  const dispatch = useDispatch();
  const inputValues = useSelector(
    (state) => state.playerCards.cardData[props.cardId].inputValues
  );

  const handleButtonClick = (dispatchFunction) => {
    dispatch(
      dispatchFunction(
        inputValues[props.labelParameter],
        props.labelParameter,
        props.cardId
      )
    );
  };

  return (
    <Grid container spacing={1}>
      <Grid item xs>
        <CardNumberInput
          labelText={props.labelText}
          minValue={1}
          maxValue={props.maxValue}
          step={props.step}
          defaultValue={props.defaultValue}
          labelParameter={props.labelParameter}
          cardId={props.cardId}
        />
      </Grid>
      <Grid item xs>
        <ButtonGroup
          color="primary"
          variant="contained"
          fullWidth={true}
          style={{ marginTop: 12 }}
        >
          <Button
            data-test="increment"
            onClick={() => handleButtonClick(increment)}
          >
            +
          </Button>
          <Button
            data-test="decrement"
            onClick={() => handleButtonClick(decrement)}
          >
            -
          </Button>
        </ButtonGroup>
      </Grid>
    </Grid>
  );
};
