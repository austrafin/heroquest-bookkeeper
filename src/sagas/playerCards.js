import { takeLatest, put } from "redux-saga/effects";
import { ADD, ADD_AFTER } from "../actions/playerCards";
import axios from "axios";

function* addPlayerCard(action) {
  yield axios
    .post("http://localhost:5000/player_cards/add", action.values)
    .catch((error) => {
      console.log(error);
    });
  yield put({ type: ADD_AFTER });
}

export const playerCardsSagas = [takeLatest(ADD, addPlayerCard)];
