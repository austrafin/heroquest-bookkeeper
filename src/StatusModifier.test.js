import React from "react";
import { shallow, mount } from "enzyme";
import StatusModifier from "./StatusModifier";
import { Provider } from "react-redux";
import { increment, decrement } from "./actions/playerCards";
import configureStore from "redux-mock-store";

jest.mock("./actions/playerCards");

describe("StatusModifier", () => {
  const mockStore = configureStore([]);
  const cardId = "test_123";
  const store = mockStore({
    playerCards: {
      cardData: {
        [cardId]: { inputValues: {} },
      },
    },
  });

  store.dispatch = jest.fn();

  beforeEach(() => {
    increment.mockClear();
    decrement.mockClear();
  });

  it("renders", () => {
    const component = shallow(
      <Provider store={store}>
        <StatusModifier />
      </Provider>
    );

    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  describe("Increment and decrement", () => {
    const component = mount(
      <Provider store={store}>
        <StatusModifier
          labelText={"test label"}
          maxValue={100}
          step={1}
          defaultValue={1}
          labelParameter="testParameter"
          cardId={cardId}
        />
      </Provider>
    );

    it("checks that the increment action is dispatched when the increment button is clicked.", () => {
      component
        .find('[data-test="increment"]')
        .find("button")
        .simulate("click");

      expect(increment).toHaveBeenCalledTimes(1);
    });

    it("checks that the increment action is dispatched when the increment button is clicked.", () => {
      component
        .find('[data-test="decrement"]')
        .find("button")
        .simulate("click");

      expect(decrement).toHaveBeenCalledTimes(1);
    });
  });
});
