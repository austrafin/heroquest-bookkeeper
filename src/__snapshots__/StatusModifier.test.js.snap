// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusModifier renders 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction] {
          "calls": Array [
            Array [
              undefined,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [MockFunction] {
            "calls": Array [
              Array [
                undefined,
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <_default />
</ContextProvider>
`;
